Kernel/Devices/KCOVDevice.cpp

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright (c) 2021, Patrick Meyer <git@the-space.agency>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Assertions.h>
+#include <AK/Format.h>
+#include <AK/NonnullOwnPtr.h>
+#include <Kernel/Devices/KCOVDevice.h>
+#include <Kernel/Devices/KCOVInstance.h>
+#include <Kernel/FileSystem/FileDescription.h>
+#include <Kernel/Sections.h>
+#include <Kernel/VM/AnonymousVMObject.h>
+#include <LibC/sys/ioctl_numbers.h>
+
+#include <Kernel/Panic.h>
+
+namespace Kernel {
+
+HashMap<ProcessID, KCOVInstance*>* KCOVDevice::proc_instance;
+HashMap<ThreadID, KCOVInstance*>* KCOVDevice::thread_instance;
+
+UNMAP_AFTER_INIT NonnullRefPtr<KCOVDevice> KCOVDevice::must_create()
+{
+    return adopt_ref_if_nonnull(new KCOVDevice).release_nonnull();
+}
+
+UNMAP_AFTER_INIT KCOVDevice::KCOVDevice()
+    : BlockDevice(30, 0)
+{
+    proc_instance = new HashMap<ProcessID, KCOVInstance*>();
+    thread_instance = new HashMap<ThreadID, KCOVInstance*>();
+    dbgln("KCOVDevice created");
+}
+
+void KCOVDevice::free_thread()
+{
+    auto thread = Thread::current();
+    auto tid = thread->tid();
+
+    auto maybe_kcov_instance = thread_instance->get(tid);
```

Comment:
The usage patterns of these static maps feels like they could be private with an api like this on the device:

```
static KCovInstance* get_instance_for_current_thread();
static KCovInstance* get_instance_for_current_process();
```
maybe?

Reply:
I would still have to handle those functions not returning a valid KCOVInstance. That is expected, as the KCOVInstance for a Process is only created when some thread in a proc first calls open() on the kcov device. If you ioctl before open those HashMaps got nothing for you. Hence I don't see how this would really simplify the code. The checks would still have to look pretty much the same, I think.

Reply:
My goal with such an API would be to hide the details of the hash map from callers. Information hiding/encapsulation reasons. 

The code that implements the getter/setter for the instances would need to still figure out if an item exists for that key, but the caller only needs to null check, and the implementation detail that they're stored in a HashMap can be hidden from the callers.

Reply:
I vote to defer this for now. I played with it a bit and I'm not too sure about the result. Let's merge this first and then discuss the change at another time.